src/csharp/Grpc.Core/ClientBase.cs

```diff
@@ -147,6 +149,64 @@ internal ClientBaseConfiguration Configuration
         /// </summary>
         protected internal class ClientBaseConfiguration
         {
+            private class ClientHeaderInterceptor : ClientInterceptor
+            {
+                readonly Func<IMethod, string, CallOptions, Tuple<string, CallOptions>> interceptor;
+
+                /// <summary>
+                /// Creates a new instance of ClientHeaderInterceptor given the specified header interceptor function.
+                /// </summary>
+                public ClientHeaderInterceptor(Func<IMethod, string, CallOptions, Tuple<string, CallOptions>> interceptor)
+                {
+                    this.interceptor = GrpcPreconditions.CheckNotNull(interceptor);
+                }
+
+                /// <summary>
+                /// Intercepts a blocking invocation of a simple remote call.
+                /// </summary>
+                public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, TResponse> next)
+                {
+                    var newHeaders = interceptor(method, host, options);
+                    return next(method, newHeaders.Item1, newHeaders.Item2, request);
+                }
+
+                /// <summary>
+                /// Intercepts an asynchronous invocation of a simple remote call.
+                /// </summary>
+                public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request, Func<Method<TRequest, TResponse>, string, CallOptions, TRequest, AsyncUnaryCall<TResponse>> next)
```

Comment:
How about `continuation`?

Reply:
continuation sounds much better.

Reply:
I have changed my mind a bit on `continuation` and have reservations about it since in the retry case, you can invoke it more than once and you can choose not to call it at all. How about `invoker`, though the downside is it is not an instance of `CallInvoker`.